### ISSUE TITLE
[1.6.0] Kotlin `suspend` functions swallow exceptions with Asynchronous execution

### ISSUE THREAD

**Issue by solonovamax** [Incendo/cloud#306]
Kotlin `suspend` functions will sometimes swallow exceptions when using `AsynchronousCommandExecutionCoordinator`

This happens because of the following:

This all happens in the `AsynchronousCommandExecutionCoordinator` class.

first, on line `117`, the `commandConsumer` is executed.
next, on line `84`, it executes the command future, but then it *doesn't return that future*
it then returns to line `117`, since that was an anonymous lambda constructed earlier, and after that it *immediately completes the future* on line `118`, so when the `whenComplete` on line `84` is later executed, and it attempts to run `completeExceptionally` on the future, it has *already completed the future*

So basically, it completes the future without waiting for the `commandConsumer`.
Which causes it to be completed *before* we know if the completion is successful or not.

So, there are two things that could be done:
1. have the `commandConsumer` throw the exception (or, just don't use `whenComplete` and actually block on the completion of the command)
2. move `resultFuture.complete` into the `commandConsumer` so we can wait for execution to finish.

Will PR with changes in a few few mins.

**Comment by TehNeon** [comment]
Mentioning my issue as this sounds either similar or identical to it, #281.

**Comment by solonovamax** [comment]
@TehNeon Were you using kotlin `suspend` functions, or no?

Also, I'm currently awaiting #307 to be merged, which will fix this.

**Comment by TehNeon** [comment]
> @TehNeon Were you using kotlin `suspend` functions, or no?
> 
> Also, I'm currently awaiting #307 to be merged, which will fix this.

I was not, but it appears from your description that it is the same/very similar issue.

**Comment by solonovamax** [comment]
Here, lemme get jitpack up for it, and can you try running with the jitpack build?

**Comment by solonovamax** [comment]
nvm, jitpack is being weird. Could you just try an artifact from building it yourself? I can upload an artifact if needed.

**Comment by solonovamax** [comment]
@TehNeon Can you test with the latest 1.6.0-dev branch?
Since #307 has been merged into 1.6.0 (not sure why it didn't close this)

**Comment by solonovamax** [comment]
Closing because it has been fixed in #307
